File: Sources/Chromabase/Export/PrintComposition.swift
import CoreGraphics
import CoreImage
import Foundation

public enum PrintPaperSize: String, CaseIterable, Codable, Sendable {
    case fourBySix
    case fiveBySeven
    case eightByTen
    case a4
    /// 사진과 같은 비율의 용지. 실제 치수는 사진의 가로세로비에서 만든다(긴 변 10 in 고정).
    case photoRatio

    /// 사진 비율 용지의 긴 변.
    public static let photoRatioLongEdgeMM: Double = 254

    public var dimensionsMM: CGSize {
        switch self {
        case .fourBySix: CGSize(width: 101.6, height: 152.4)
        case .fiveBySeven: CGSize(width: 127, height: 177.8)
        case .eightByTen: CGSize(width: 203.2, height: 254)
        case .a4: CGSize(width: 210, height: 297)
        case .photoRatio: CGSize(width: Self.photoRatioLongEdgeMM * 2 / 3, height: Self.photoRatioLongEdgeMM)
        }
    }

    /// 사진 비율(가로/세로)에서 용지 치수를 만든다. 비율을 모르면 3:2 로 둔다.
    public func dimensionsMM(photoAspectRatio: Double?) -> CGSize {
        guard self == .photoRatio,
              let photoAspectRatio,
              photoAspectRatio.isFinite,
              photoAspectRatio > 0 else { return dimensionsMM }
        let longEdge = Self.photoRatioLongEdgeMM
        return photoAspectRatio >= 1
            ? CGSize(width: longEdge, height: longEdge / photoAspectRatio)
            : CGSize(width: longEdge * photoAspectRatio, height: longEdge)
    }

    public var uiLabel: String {
        switch self {
        case .fourBySix: "4 × 6 in"
        case .fiveBySeven: "5 × 7 in"
        case .eightByTen: "8 × 10 in"
        case .a4: "A4"
        case .photoRatio: "Photo"
        }
    }
}

public enum PrintPaperOrientation: String, CaseIterable, Codable, Sendable {
    case automatic
    case portrait
    case landscape
}

public enum PrintPerforationStyle: String, CaseIterable, Codable, Sendable {
    case none
    case thirtyFiveMillimeter
}

/// 인화 레이아웃이 사진에 적용하는 고정 표현 방식.
///
/// 측정 프로파일을 대신하는 장치 정확도 시뮬레이션이 아니라, 공정의 핵심 시각 특성만
/// 화면과 파일에 동일하게 적용한다.
public enum PrintPresentationStyle: String, CaseIterable, Codable, Sendable {
    case standard
    case cyanotype
    case glassPlate
    case gelatinSilver
}

public struct PrintPresentationAppearance: Equatable, Sendable {
    public let shadowRGBA: SIMD4<Double>
    public let highlightRGBA: SIMD4<Double>

    public init(style: PrintPresentationStyle) {
        switch style {
        case .standard:
            shadowRGBA = SIMD4(0, 0, 0, 1)
            highlightRGBA = SIMD4(1, 1, 1, 1)
        case .cyanotype:
            // 시아노타입의 철염 이미지가 갖는 청색 단색 관계만 표현한다.
            shadowRGBA = SIMD4(0.02, 0.10, 0.36, 1)
            highlightRGBA = SIMD4(0.96, 0.98, 1, 1)
        case .glassPlate:
            shadowRGBA = SIMD4(0, 0, 0, 1)
            highlightRGBA = SIMD4(1, 1, 1, 1)
        case .gelatinSilver:
            shadowRGBA = SIMD4(0, 0, 0, 1)
            highlightRGBA = SIMD4(1, 1, 1, 1)
        }
    }
}

public struct PrintFilmStripAppearance: Equatable, Sendable {
    public let baseRGBA: SIMD4<Double>

    public init(filmType: FilmType) {
        switch filmType {
        case .colorNegative:
            // 현상된 컬러 네거티브의 마스크가 남은 비노광 가장자리.
            baseRGBA = SIMD4(0.44, 0.17, 0.055, 0.92)
        case .bwNegative:
            // 흑백 네거티브의 중성 회색 투명 베이스.
            baseRGBA = SIMD4(0.12, 0.12, 0.12, 0.84)
        case .colorPositive:
            // 컬러 리버설의 비노광 가장자리는 거의 불투명한 흑색이다.
            baseRGBA = SIMD4(0.025, 0.032, 0.042, 0.97)
        case .bwPositive:
            // 흑백 리버설의 비노광 가장자리는 중성 흑색이다.
            baseRGBA = SIMD4(0.022, 0.022, 0.022, 0.97)
        }
    }
}

public struct PrintCompositionSettings: Codable, Equatable, Sendable {
    public var paperSize: PrintPaperSize
    public var orientation: PrintPaperOrientation
    public var marginMM: Double
    public var dpi: Int
    public var perforationStyle: PrintPerforationStyle
    /// `photoRatio` 용지가 따라갈 사진의 가로/세로비. 다른 용지에서는 무시된다.
    public var photoAspectRatio: Double?
    public var presentationStyle: PrintPresentationStyle

    public init(
        paperSize: PrintPaperSize = .a4,
        orientation: PrintPaperOrientation = .automatic,
        marginMM: Double = 10,
        dpi: Int = 300,
        perforationStyle: PrintPerforationStyle = .none,
        photoAspectRatio: Double? = nil,
        presentationStyle: PrintPresentationStyle = .standard
    ) {
        self.paperSize = paperSize
        self.orientation = orientation
        self.marginMM = marginMM
        self.dpi = dpi
        self.perforationStyle = perforationStyle
        self.photoAspectRatio = photoAspectRatio
        self.presentationStyle = presentationStyle
    }

    /// 이 설정이 쓰는 실제 용지 치수.
    public var paperDimensionsMM: CGSize {
        paperSize.dimensionsMM(photoAspectRatio: photoAspectRatio)
    }

    public var isValid: Bool {
        marginMM.isFinite && (0...50).contains(marginMM) && (72...600).contains(dpi)
    }

    private enum CodingKeys: String, CodingKey {
        case paperSize
        case orientation
        case marginMM
        case dpi
        case perforationStyle
        case photoAspectRatio
        case presentationStyle
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        paperSize = try container.decode(PrintPaperSize.self, forKey: .paperSize)
        orientation = try container.decode(PrintPaperOrientation.self, forKey: .orientation)
        marginMM = try container.decode(Double.self, forKey: .marginMM)
        dpi = try container.decode(Int.self, forKey: .dpi)
        perforationStyle = try container.decode(
            PrintPerforationStyle.self,
            forKey: .perforationStyle
        )
        photoAspectRatio = try container.decodeIfPresent(Double.self, forKey: .photoAspectRatio)
        presentationStyle = try container.decodeIfPresent(
            PrintPresentationStyle.self,
            forKey: .presentationStyle
        ) ?? .standard
    }
}

public struct PrintCompositionLayout: Equatable, Sendable {
    public let canvasSize: CGSize
    public let contentRect: CGRect
    public let imageRect: CGRect
    public let filmRect: CGRect?
    public let perforationRects: [CGRect]
    public let perforationCornerRadius: CGFloat

    public static func make(
        sourceSize: CGSize,
        settings: PrintCompositionSettings
    ) -> PrintCompositionLayout? {
        guard sourceSize.width > 0,
              sourceSize.height > 0,
              sourceSize.width.isFinite,
              sourceSize.height.isFinite,
              settings.isValid else { return nil }

        let sourceIsLandscape = sourceSize.width >= sourceSize.height
        let base = settings.paperDimensionsMM
        let pageMM: CGSize
        switch settings.orientation {
        case .automatic:
            pageMM = sourceIsLandscape
                ? CGSize(width: max(base.width, base.height), height: min(base.width, base.height))
                : CGSize(width: min(base.width, base.height), height: max(base.width, base.height))
        case .portrait:
            pageMM = CGSize(width: min(base.width, base.height), height: max(base.width, base.height))
        case .landscape:
            pageMM = CGSize(width: max(base.width, base.height), height: min(base.width, base.height))
        }

        let pixelsPerMM = CGFloat(settings.dpi) / 25.4
        let canvasSize = CGSize(
            width: max(1, (pageMM.width * pixelsPerMM).rounded()),
            height: max(1, (pageMM.height * pixelsPerMM).rounded())
        )
        let margin = CGFloat(settings.marginMM) * pixelsPerMM
        let contentRect = CGRect(origin: .zero, size: canvasSize).insetBy(dx: margin, dy: margin)
        guard contentRect.width > 1, contentRect.height > 1 else { return nil }

        if settings.perforationStyle == .none {
            return PrintCompositionLayout(
                canvasSize: canvasSize,
                contentRect: contentRect,
                imageRect: aspectFit(sourceSize, in: contentRect),
                filmRect: nil,
                perforationRects: [],
                perforationCornerRadius: 0
            )
        }

        let isLandscape = sourceSize.width >= sourceSize.height
        // ISO 1007의 135 풀프레임 기준: 35 mm 폭, 24 × 36 mm 이미지 게이트,
        // 프레임 피치 38 mm(4.75 mm KS-1870 천공 8개).
        let physicalFilmSizeMM = isLandscape
            ? CGSize(width: 38, height: 35)
            : CGSize(width: 35, height: 38)
        let filmRect = aspectFit(physicalFilmSizeMM, in: contentRect)
        let unit = isLandscape ? filmRect.height / 35 : filmRect.width / 35
        let gateSizeMM = isLandscape
            ? CGSize(width: 36, height: 24)
            : CGSize(width: 24, height: 36)
        let gateRect = CGRect(
            x: filmRect.midX - gateSizeMM.width * unit / 2,
            y: filmRect.midY - gateSizeMM.height * unit / 2,
            width: gateSizeMM.width * unit,
            height: gateSizeMM.height * unit
        )
        let imageRect = aspectFit(sourceSize, in: gateRect)

        let pitch = 4.75 * unit
        let railCenterOffset = 2.75 * unit
        var perforationRects: [CGRect] = []
        perforationRects.reserveCapacity(16)
        if isLandscape {
            let holeSize = CGSize(width: 2.79 * unit, height: 1.98 * unit)
            let occupiedWidth = 7 * pitch + holeSize.width
            let firstX = filmRect.midX - occupiedWidth / 2
            let bottomY = filmRect.minY + railCenterOffset - holeSize.height / 2
            let topY = filmRect.maxY - railCenterOffset - holeSize.height / 2
            for index in 0..<8 {
                let x = firstX + CGFloat(index) * pitch
                perforationRects.append(CGRect(origin: CGPoint(x: x, y: bottomY), size: holeSize))
                perforationRects.append(CGRect(origin: CGPoint(x: x, y: topY), size: holeSize))
            }
        } else {
            let holeSize = CGSize(width: 1.98 * unit, height: 2.79 * unit)
            let occupiedHeight = 7 * pitch + holeSize.height
            let firstY = filmRect.midY - occupiedHeight / 2
            let leftX = filmRect.minX + railCenterOffset - holeSize.width / 2
            let rightX = filmRect.maxX - railCenterOffset - holeSize.width / 2
            for index in 0..<8 {
                let y = firstY + CGFloat(index) * pitch
                perforationRects.append(CGRect(origin: CGPoint(x: leftX, y: y), size: holeSize))
                perforationRects.append(CGRect(origin: CGPoint(x: rightX, y: y), size: holeSize))
            }
        }

        return PrintCompositionLayout(
            canvasSize: canvasSize,
            contentRect: contentRect,
            imageRect: imageRect,
            filmRect: filmRect,
            perforationRects: perforationRects,
            perforationCornerRadius: 0.51 * unit
        )
    }

    private static func aspectFit(_ size: CGSize, in bounds: CGRect) -> CGRect {
        let scale = min(bounds.width / size.width, bounds.height / size.height)
        let fitted = CGSize(width: size.width * scale, height: size.height * scale)
        return CGRect(
            x: bounds.midX - fitted.width / 2,
            y: bounds.midY - fitted.height / 2,
            width: fitted.width,
            height: fitted.height
        )
    }
}

public enum PrintCompositionRenderer {
    public static func apply(
        to image: CIImage,
        settings: PrintCompositionSettings,
        filmType: FilmType = .colorPositive
    ) -> CIImage? {
        guard let layout = PrintCompositionLayout.make(
            sourceSize: image.extent.size,
            settings: settings
        ) else { return nil }

        let canvasRect = CGRect(origin: .zero, size: layout.canvasSize)
        let paper = CIImage(color: CIColor(red: 1, green: 1, blue: 1, alpha: 1))
            .cropped(to: canvasRect)
        var result = paper

        if let filmRect = layout.filmRect {
            let rgba = PrintFilmStripAppearance(filmType: filmType).baseRGBA
            let film = CIImage(color: CIColor(
                red: CGFloat(rgba.x),
                green: CGFloat(rgba.y),
                blue: CGFloat(rgba.z),
                alpha: CGFloat(rgba.w)
            ))
                .cropped(to: filmRect)
            result = film.composited(over: result)
        }

        let presented = PrintPresentationRenderer.apply(
            to: image,
            style: settings.presentationStyle
        )
        let placed = place(presented, in: layout.imageRect)
        result = placed.composited(over: result)

        for rect in layout.perforationRects {
            guard let hole = roundedRectangle(
                rect: rect,
                radius: layout.perforationCornerRadius,
                color: CIColor(red: 1, green: 1, blue: 1, alpha: 1)
            ) else { return nil }
            result = hole.composited(over: result)
        }
        return result.cropped(to: canvasRect)
    }

    private static func roundedRectangle(
        rect: CGRect,
        radius: CGFloat,
        color: CIColor
    ) -> CIImage? {
        CIFilter(
            name: "CIRoundedRectangleGenerator",
            parameters: [
                "inputExtent": CIVector(cgRect: rect),
                "inputRadius": radius,
                kCIInputColorKey: color
            ]
        )?.outputImage?.cropped(to: rect)
    }

    private static func place(_ image: CIImage, in rect: CGRect) -> CIImage {
        let normalized = normalize(image)
        let scale = min(rect.width / normalized.extent.width, rect.height / normalized.extent.height)
        let scaled = normalized.transformed(by: CGAffineTransform(scaleX: scale, y: scale))
        return scaled.transformed(by: CGAffineTransform(
            translationX: rect.midX - scaled.extent.midX,
            y: rect.midY - scaled.extent.midY
        ))
    }

    private static func normalize(_ image: CIImage) -> CIImage {
        let extent = image.extent
        guard extent.origin != .zero else { return image }
        return image.transformed(by: CGAffineTransform(
            translationX: -extent.minX,
            y: -extent.minY
        ))
    }
}

public enum PrintPresentationRenderer {
    /// 필터 그래프만 연결하고 여기서는 픽셀을 렌더하지 않는다. 최종 export의 공유 `CIContext`가
    /// 한 번에 평가하므로 중간 비트맵과 추가 디코드를 만들지 않는다.
    public static func apply(
        to image: CIImage,
        style: PrintPresentationStyle
    ) -> CIImage {
        switch style {
        case .standard:
            return image
        case .cyanotype:
            let palette = PrintPresentationAppearance(style: style)
            return monochrome(image).applyingFilter(
                "CIFalseColor",
                parameters: [
                    "inputColor0": ciColor(palette.shadowRGBA),
                    "inputColor1": ciColor(palette.highlightRGBA),
                ]
            )
        case .glassPlate:
            return monochrome(image).applyingFilter("CIColorInvert")
        case .gelatinSilver:
            return monochrome(image)
        }
    }

    private static func monochrome(_ image: CIImage) -> CIImage {
        image.applyingFilter(
            "CIColorControls",
            parameters: [kCIInputSaturationKey: 0]
        )
    }

    private static func ciColor(_ rgba: SIMD4<Double>) -> CIColor {
        CIColor(
            red: CGFloat(rgba.x),
            green: CGFloat(rgba.y),
            blue: CGFloat(rgba.z),
            alpha: CGFloat(rgba.w)
        )
    }
}

File: negaflow-windows/src/Shell.Core/Print/PrintComposition.cs
namespace Negaflow.Shell.Print;

/// <summary>용지입니다. macOS <c>PrintPaperSize</c> 와 같은 스물여섯 가지, 같은 차례입니다.</summary>
public enum PrintPaperSize
{
    PhotoRatio,
    ThreePointFiveByFive,
    FourBySix,
    FiveBySeven,
    EightByTen,
    TenByTwelve,
    ElevenByFourteen,
    TwelveByEighteen,
    SixteenByTwenty,
    TwentyByTwentyFour,
    TwentyByThirty,
    TwentyFourByThirtySix,
    Letter,
    Tabloid,
    A3Plus,
    A6,
    A5,
    A4,
    A3,
    A2,
    A1,
    B6,
    B5,
    B4,
    B3,
    B2,
    B1,
}

public enum PrintPaperOrientation
{
    Automatic,
    Portrait,
    Landscape,
}

public enum PrintPerforationStyle
{
    None,
    ThirtyFiveMillimeter,
}

/// <summary>
/// 인화 공정의 겉모습입니다. 측정 프로파일을 대신하는 장치 정확도 시뮬레이션이 아니라, 공정의
/// 핵심 시각 특성만 화면과 파일에 <b>같게</b> 적용합니다 — macOS 와 같은 뜻입니다.
/// </summary>
public enum PrintPresentationStyle
{
    Standard,
    Cyanotype,
    GlassPlate,
    GelatinSilver,
}

public enum PrintSheetBackground
{
    Black,
    Gray,
    White,
}

public enum PrintLayoutMode
{
    SingleImage,
    ContactSheet,
    PicturePackage,
    CustomPackage,
    Cyanotype,
    GlassPlate,
    Gelatin,
}

public enum PrintPaperSurface
{
    Glossy,
    Matte,
    Lustre,
    Silk,
}

public enum PrintRulerUnit
{
    Inches,
    Centimeters,
}

public enum PrintOutputProcess
{
    Standard,
    CPrint,
}

public readonly record struct PrintSizeMm(double Width, double Height);

public readonly record struct PrintRect(double X, double Y, double Width, double Height)
{
    public double MinX => X;

    public double MinY => Y;

    public double MaxX => X + Width;

    public double MaxY => Y + Height;

    public double MidX => X + (Width / 2);

    public double MidY => Y + (Height / 2);

    /// <summary>안쪽으로 물러난 사각형입니다. 남는 것이 없으면 폭·높이가 음수가 됩니다.</summary>
    public PrintRect Inset(double amount) =>
        new(X + amount, Y + amount, Width - (amount * 2), Height - (amount * 2));
}

public static class PrintPaper
{
    /// <summary>사진 비율 용지의 긴 변입니다. macOS 와 같은 254mm 입니다.</summary>
    public const double PhotoRatioLongEdgeMm = 254;

    public static IReadOnlyList<PrintPaperSize> All { get; } =
        [.. Enum.GetValues<PrintPaperSize>()];

    /// <summary>mm 치수입니다. macOS <c>dimensionsMM</c> 과 <b>같은 수</b>여야 합니다.</summary>
    public static PrintSizeMm DimensionsMm(PrintPaperSize size) => size switch
    {
        PrintPaperSize.PhotoRatio =>
            new(PhotoRatioLongEdgeMm * 2 / 3, PhotoRatioLongEdgeMm),
        PrintPaperSize.ThreePointFiveByFive => new(88.9, 127),
        PrintPaperSize.FourBySix => new(101.6, 152.4),
        PrintPaperSize.FiveBySeven => new(127, 177.8),
        PrintPaperSize.EightByTen => new(203.2, 254),
        PrintPaperSize.TenByTwelve => new(254, 304.8),
        PrintPaperSize.ElevenByFourteen => new(279.4, 355.6),
        PrintPaperSize.TwelveByEighteen => new(304.8, 457.2),
        PrintPaperSize.SixteenByTwenty => new(406.4, 508),
        PrintPaperSize.TwentyByTwentyFour => new(508, 609.6),
        PrintPaperSize.TwentyByThirty => new(508, 762),
        PrintPaperSize.TwentyFourByThirtySix => new(609.6, 914.4),
        PrintPaperSize.Letter => new(215.9, 279.4),
        PrintPaperSize.Tabloid => new(279.4, 431.8),
        PrintPaperSize.A3Plus => new(329, 483),
        PrintPaperSize.A6 => new(105, 148),
        PrintPaperSize.A5 => new(148, 210),
        PrintPaperSize.A4 => new(210, 297),
        PrintPaperSize.A3 => new(297, 420),
        PrintPaperSize.A2 => new(420, 594),
        PrintPaperSize.A1 => new(594, 841),
        PrintPaperSize.B6 => new(125, 176),
        PrintPaperSize.B5 => new(176, 250),
        PrintPaperSize.B4 => new(250, 353),
        PrintPaperSize.B3 => new(353, 500),
        PrintPaperSize.B2 => new(500, 707),
        _ => new(707, 1000),
    };

    /// <summary>
    /// 사진 비율 용지는 사진을 따라갑니다. 비율을 모르면 macOS 처럼 3:2 로 둡니다.
    /// </summary>
    public static PrintSizeMm DimensionsMm(PrintPaperSize size, double? photoAspectRatio)
    {
        if (size != PrintPaperSize.PhotoRatio ||
            photoAspectRatio is not { } ratio ||
            !double.IsFinite(ratio) ||
            ratio <= 0)
        {
            return DimensionsMm(size);
        }
        return ratio >= 1
            ? new PrintSizeMm(PhotoRatioLongEdgeMm, PhotoRatioLongEdgeMm / ratio)
            : new PrintSizeMm(PhotoRatioLongEdgeMm * ratio, PhotoRatioLongEdgeMm);
    }

    /// <summary>
    /// 고르개에 보이는 이름입니다. 번역하지 않습니다 — A4·Letter 는 규격 이름이고 치수 표기는
    /// 언어와 무관합니다. macOS <c>uiLabel</c> 과 같습니다.
    /// </summary>
    public static string Label(PrintPaperSize size) => size switch
    {
        PrintPaperSize.PhotoRatio => "Photo",
        PrintPaperSize.ThreePointFiveByFive => "3.5 × 5 in",
        PrintPaperSize.FourBySix => "4 × 6 in",
        PrintPaperSize.FiveBySeven => "5 × 7 in",
        PrintPaperSize.EightByTen => "8 × 10 in",
        PrintPaperSize.TenByTwelve => "10 × 12 in",
        PrintPaperSize.ElevenByFourteen => "11 × 14 in",
        PrintPaperSize.TwelveByEighteen => "12 × 18 in",
        PrintPaperSize.SixteenByTwenty => "16 × 20 in",
        PrintPaperSize.TwentyByTwentyFour => "20 × 24 in",
        PrintPaperSize.TwentyByThirty => "20 × 30 in",
        PrintPaperSize.TwentyFourByThirtySix => "24 × 36 in",
        PrintPaperSize.Letter => "Letter · 8.5 × 11 in",
        PrintPaperSize.Tabloid => "Tabloid · 11 × 17 in",
        PrintPaperSize.A3Plus => "A3+ · 13 × 19 in",
        _ => size.ToString().ToUpperInvariant(),
    };
}

/// <summary>인화 판 하나를 정하는 값입니다. macOS <c>PrintCompositionSettings</c> 과 같습니다.</summary>
public sealed record PrintCompositionSettings
{
    public PrintPaperSize PaperSize { get; init; } = PrintPaperSize.A4;

    public PrintPaperOrientation Orientation { get; init; } = PrintPaperOrientation.Automatic;

    public double MarginMm { get; init; } = 10;

    public int Dpi { get; init; } = 300;

    public PrintPerforationStyle PerforationStyle { get; init; } = PrintPerforationStyle.None;

    /// <summary><c>PhotoRatio</c> 용지가 따라갈 사진의 가로/세로비입니다. 다른 용지는 무시합니다.</summary>
    public double? PhotoAspectRatio { get; init; }

    public PrintPresentationStyle PresentationStyle { get; init; } =
        PrintPresentationStyle.Standard;

    public PrintSheetBackground SheetBackground { get; init; } = PrintSheetBackground.White;

    public PrintSizeMm PaperDimensionsMm => PrintPaper.DimensionsMm(PaperSize, PhotoAspectRatio);

    /// <summary>macOS 와 같은 한계입니다 — 여백 0~50mm, 해상도 72~600dpi.</summary>
    public bool IsValid =>
        double.IsFinite(MarginMm) && MarginMm is >= 0 and <= 50 && Dpi is >= 72 and <= 600;
}

/// <summary>
/// 한 장을 어떻게 놓을지 계산한 결과입니다. 화면 미리보기와 내보내기가 <b>같은 값</b>을 씁니다 —
/// 둘이 다른 계산을 하면 보이는 것과 나오는 것이 갈립니다.
/// </summary>
public sealed record PrintCompositionLayout(
    PrintSizeMm CanvasSize,
    PrintRect ContentRect,
    PrintRect ImageRect,
    PrintRect? FilmRect,
    IReadOnlyList<PrintRect> PerforationRects,
    double PerforationCornerRadius)
{
    /// <summary>
    /// 계산합니다. 사진 크기나 설정이 말이 안 되면 null 입니다 — 억지로 한 장을 만들면 사용자는
    /// 왜 이상한 판이 나왔는지 알 수 없습니다.
    /// </summary>
    public static PrintCompositionLayout? Make(
        PrintSizeMm sourceSize,
        PrintCompositionSettings settings)
    {
        ArgumentNullException.ThrowIfNull(settings);
        if (sourceSize.Width <= 0 || sourceSize.Height <= 0 ||
            !double.IsFinite(sourceSize.Width) || !double.IsFinite(sourceSize.Height) ||
            !settings.IsValid)
        {
            return null;
        }

        bool sourceIsLandscape = sourceSize.Width >= sourceSize.Height;
        PrintSizeMm page = PageDimensions(
            settings.PaperDimensionsMm,
            settings.Orientation,
            sourceIsLandscape);

        double pixelsPerMm = settings.Dpi / 25.4;
        PrintSizeMm canvas = new(
            Math.Max(1, Math.Round(page.Width * pixelsPerMm)),
            Math.Max(1, Math.Round(page.Height * pixelsPerMm)));
        double margin = settings.MarginMm * pixelsPerMm;
        PrintRect content = new PrintRect(0, 0, canvas.Width, canvas.Height).Inset(margin);
        if (content.Width <= 1 || content.Height <= 1)
        {
            return null;
        }

        if (settings.PerforationStyle == PrintPerforationStyle.None)
        {
            return new PrintCompositionLayout(
                canvas,
                content,
                AspectFit(sourceSize, content),
                null,
                [],
                0);
        }

        // ISO 1007 의 135 풀프레임 기준입니다 — 35mm 폭, 24×36mm 이미지 게이트, 프레임 피치
        // 38mm(4.75mm KS-1870 천공 8개). macOS 와 같은 수입니다.
        PrintSizeMm filmMm = sourceIsLandscape ? new(38, 35) : new(35, 38);
        PrintRect film = AspectFit(filmMm, content);
        double unit = sourceIsLandscape ? film.Height / 35 : film.Width / 35;
        PrintSizeMm gateMm = sourceIsLandscape ? new(36, 24) : new(24, 36);
        PrintRect gate = new(
            film.MidX - (gateMm.Width * unit / 2),
            film.MidY - (gateMm.Height * unit / 2),
            gateMm.Width * unit,
            gateMm.Height * unit);

        double pitch = 4.75 * unit;
        double railCenterOffset = 2.75 * unit;
        List<PrintRect> perforations = new(16);
        if (sourceIsLandscape)
        {
            PrintSizeMm hole = new(2.79 * unit, 1.98 * unit);
            double firstX = film.MidX - (((7 * pitch) + hole.Width) / 2);
            double bottomY = film.MinY + railCenterOffset - (hole.Height / 2);
            double topY = film.MaxY - railCenterOffset - (hole.Height / 2);
            for (int index = 0; index < 8; ++index)
            {
                double x = firstX + (index * pitch);
                perforations.Add(new PrintRect(x, bottomY, hole.Width, hole.Height));
                perforations.Add(new PrintRect(x, topY, hole.Width, hole.Height));
            }
        }
        else
        {
            PrintSizeMm hole = new(1.98 * unit, 2.79 * unit);
            double firstY = film.MidY - (((7 * pitch) + hole.Height) / 2);
            double leftX = film.MinX + railCenterOffset - (hole.Width / 2);
            double rightX = film.MaxX - railCenterOffset - (hole.Width / 2);
            for (int index = 0; index < 8; ++index)
            {
                double y = firstY + (index * pitch);
                perforations.Add(new PrintRect(leftX, y, hole.Width, hole.Height));
                perforations.Add(new PrintRect(rightX, y, hole.Width, hole.Height));
            }
        }

        return new PrintCompositionLayout(
            canvas,
            content,
            AspectFit(sourceSize, gate),
            film,
            perforations,
            0.51 * unit);
    }

    /// <summary>
    /// 방향을 정한 뒤의 용지 치수입니다. 자동은 사진을 따라갑니다 — 가로 사진에 세로 용지를
    /// 내밀면 사용자가 매번 방향을 고쳐야 합니다.
    /// </summary>
    internal static PrintSizeMm PageDimensions(
        PrintSizeMm paper,
        PrintPaperOrientation orientation,
        bool sourceIsLandscape)
    {
        double longEdge = Math.Max(paper.Width, paper.Height);
        double shortEdge = Math.Min(paper.Width, paper.Height);
        bool landscape = orientation switch
        {
            PrintPaperOrientation.Portrait => false,
            PrintPaperOrientation.Landscape => true,
            _ => sourceIsLandscape,
        };
        return landscape ? new PrintSizeMm(longEdge, shortEdge) : new PrintSizeMm(shortEdge, longEdge);
    }

    /// <summary>비율을 지키며 가운데 맞춰 넣습니다.</summary>
    internal static PrintRect AspectFit(PrintSizeMm size, PrintRect bounds)
    {
        double scale = Math.Min(bounds.Width / size.Width, bounds.Height / size.Height);
        double width = size.Width * scale;
        double height = size.Height * scale;
        return new PrintRect(
            bounds.MidX - (width / 2),
            bounds.MidY - (height / 2),
            width,
            height);
    }
}

/// <summary>
/// 공정별 겉모습입니다. 그림자와 하이라이트가 어느 색으로 가는지만 정합니다 — macOS
/// <c>PrintPresentationAppearance</c> 와 같은 수입니다.
/// </summary>
public readonly record struct PrintPresentationAppearance(
    double ShadowRed,
    double ShadowGreen,
    double ShadowBlue,
    double HighlightRed,
    double HighlightGreen,
    double HighlightBlue)
{
    public static PrintPresentationAppearance For(PrintPresentationStyle style) => style switch
    {
        // 시아노타입의 철염 이미지가 갖는 청색 단색 관계만 표현합니다.
        PrintPresentationStyle.Cyanotype => new(0.02, 0.10, 0.36, 0.96, 0.98, 1),
        _ => new(0, 0, 0, 1, 1, 1),
    };
}
